dftimewolf/lib/processors/gcp_cloud_resource_tree.py

```diff
@@ -0,0 +1,1002 @@
+# -*- coding: utf-8 -*-
+"""Generates a GCP cloud resource tree."""
+
+import json
+import tempfile
+from datetime import datetime
+from datetime import timedelta
+from datetime import timezone
+from typing import Dict, List, Optional, Any
+
+from libcloudforensics.providers.gcp.internal.common import CreateService
+from libcloudforensics.providers.gcp.internal.log import GoogleCloudLog
+
+from dftimewolf.lib import module
+from dftimewolf.lib.containers import containers
+from dftimewolf.lib.modules import manager as modules_manager
+from dftimewolf.lib.state import DFTimewolfState
+
+from dftimewolf.lib.processors.gcp_cloud_resource_tree_helper import OperatingMode, Resource  # pylint: disable=line-too-long
+
+
+class GCPCloudResourceTree(module.BaseModule):
+  """GCP Cloud Resource Tree Creator.
+
+  Attributes:
+    project_id (str): Id of the project where the resource is located.
+    resource_name (str): Name of the resource to build the tree for.
+    resource_type (str): Resource type.
+    resources_dict (Dict[str, Resource]): Dictionary of resources
+    period_covered_by_retrieved_logs (Dict[str, datetime]): Dictionary of the time
+        range of logs retrieved and processed
+
+  """
+
+  def __init__(self,
+               state: DFTimewolfState,
+               name: Optional[str] = None,
+               critical: bool = True) -> None:
+    """Initializes the Cloud Resource Tree Processor.
+
+    Args:
+      state: recipe state.
+      name: The module's runtime name.
+      critical: True if the module is critical, which causes
+          the entire recipe to fail if the module encounters an error.
+    """
+    super(GCPCloudResourceTree, self).__init__(state,
+                                               name=name,
+                                               critical=critical)
+
+    self.project_id: str = ''
+    self.resource_name: str = ''
+    self.resource_type: str = ''
+    self.mode: OperatingMode = OperatingMode.ONLINE
+    self.period_covered_by_retrieved_logs: Dict[str, datetime] = {}
+    self.resources_dict: Dict[str, Resource] = {}
+
+  # pylint: disable=arguments-differ
+  def SetUp(self, project_id: str, resource_name: str, resource_type: str,
+            mode: str) -> None:
+    """Sets up the resource we want to build the tree for.
+
+    Args:
+      project_id: Project id where the resources are located.
+      resource_name: Resource name.
+      resource_type: Resource type (currently supported types: gce_instance,
+          gce_disk, gce_image, gce_machine_image, gce_instance_template,
+          gce_snapshot)
+      mode: operational mode (online or offline)
+    """
+    self.project_id = project_id
+    self.resource_name = resource_name
+    self.resource_type = resource_type
+    if 'offline' in mode:
+      self.mode = OperatingMode.OFFLINE
+    elif 'online' in mode:
+      self.mode = OperatingMode.ONLINE
```

Comment:
if mode is not offline or online you should raise an error and quite here and not in process, i.e. fail early 
dftimewolf try to do all the quick verifications in the setup method, setup methods from all modules in a recipe are run first, the process method will only be run after all setup methods of all modules and all process methods of the prior modules are run which can lead to waiting for 5 mins and then getting a message "invalid mode"

Reply:
Done